### ISSUE TITLE
TypeScript class extending Set returns error.

### ISSUE THREAD

**Issue by SimonKovac** [Microsoft/ChakraCore#5491]
When trying to extend a set to allow unique custom objects, I found an error stating:
`Set: 'this' is not a Set object`

Recreation:
https://jsfiddle.net/9xknubyj/

**Comment by MSLaguana** [comment]
I just tried running that jsfiddle, and I didn't get any errors. If I add the lines

```js
var m = new MySet<string>();
m.add("foo");
```

then I do get an error, but the error is to do with the use of `super` (or more specifically, `_super` once converted to javascript) in the `add` method, which looks correct to me. Where were you seeing the error you report? What version of Edge are you using, and does the error show up in the console when running the exact jsfiddle you link to?

**Comment by SimonKovac** [comment]
I got the error when running the exact piece of code in a separate project, or when running it through the typescript playground(http://www.typescriptlang.org/play/). The error doesn't show up on IE11, which is why I posted the issue here. 
Edge version is 15, the error you got is showing up in the console of the Jsfiddle, but I'm not sure what kind of compilers they are using.

**Comment by MSLaguana** [comment]
Ah I see; using the typescript playground I do get the error that you report. However, I also get a similar bug if I try to repeat the same thing using v8 (using node 10.6.0 as a host):

> TypeError: Constructor Set requires 'new'

Do you see this code working as you expect anywhere? This may be a bug in the way typescript is converting to javascript?

**Comment by MSLaguana** [comment]
Also to clarify: you say that it works in ie11, but I believe that will be because ie11 doesn't have full ES6 support, and so ie11 doesn't have the notion of classes. Edge does have ES6 support which changes the behavior of things like `Set` and `Map`, and that may be what is causing the discrepancy here.

**Comment by fatcerberus** [comment]
The same error occurs when using the TypeScript Playground in Chrome.  The behavior is unfortunately correct per the ECMAScript specification.

The issue appears to be that the transpiled constructor does:
```js
_super.apply(this, arguments)
```

This is not legal for class constructors (they **must** be called using either `new` or `Reflect.construct()`). 
 Setting the TS compiler to target ES6 or higher would probably fix it.  Sadly the Playground doesn't seem to have an option to do that.

Long story short: ES5-ified classes cannot extend native ES6 classes.

See also: https://github.com/babel/babel/issues/6299 (this issue bit me in Babel too)

**Comment by kfarnung** [comment]
If I run the code sample using `tsc` set to `es5` mode then I get the following build errors:

```console
> npx tsc
src/index.ts:16:40 - error TS2304: Cannot find name 'Set'.

16 class MySet<T extends SetItem> extends Set<T> {
                                          ~~~


src/index.ts:19:14 - error TS2339: Property 'forEach' does not exist on type 'MySet<T>'.

19         this.forEach(item => {
                ~~~~~~~


src/index.ts:19:22 - error TS7006: Parameter 'item' implicitly has an 'any' type.

19         this.forEach(item => {
                        ~~~~

```

If I switch my `tsconfig.json` to `es2015` then I get correct output which uses class syntax.  Since Set wasn't introduced until ES2015, you'll likely need to take another approach.
